File: SudokuSolver_Java_No_Streams/src/main/sudokusolver/javanostreams/logic/simple/NakedSingles.java
package sudokusolver.javanostreams.logic.simple;

import sudokusolver.javanostreams.Board;
import sudokusolver.javanostreams.Cell;
import sudokusolver.javanostreams.SetValue;
import sudokusolver.javanostreams.UnsolvedCell;

import java.util.ArrayList;
import java.util.List;

/*
 * https://www.sudokuwiki.org/Getting_Started
 *
 * If an unsolved cell has exactly one candidate, then the candidate must be placed in that cell.
 */
public class NakedSingles {
    public static List<SetValue> nakedSingles(Board<Cell> board) {
        var modifications = new ArrayList<SetValue>();
        for (var cell : board.getCells()) {
            if (cell instanceof UnsolvedCell unsolved && unsolved.candidates().size() == 1) {
                modifications.add(new SetValue(unsolved, unsolved.candidates().iterator().next()));
            }
        }
        return modifications;
    }
}

File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/NakedSingles.kt
package sudokusolver.kotlin.logic

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.SetValue
import sudokusolver.kotlin.UnsolvedCell

fun nakedSingles(board: Board<Cell>): List<SetValue> =
    board.cells
        .filterIsInstance<UnsolvedCell>()
        .filter { it.candidates.size == 1 }
        .map { SetValue(it, it.candidates.first()) }